src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,90 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Diagnostics.Tracing;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal class EventCountersMetrics : EventListener
+{
+    internal static readonly AssemblyName AssemblyName = typeof(EventCountersMetrics).Assembly.GetName();
+    internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentBag<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, Instrument> instruments = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
+
+        while (this.preInitEventSources.TryTake(out EventSource source))
+        {
+            if (this.options.Sources.Contains(source.Name))
+            {
+                this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, this.options.EnableEventsArguments);
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Add(source);
+        }
+        else if (this.options.Sources.Contains(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, this.options.EnableEventsArguments);
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null || !this.options.Names.Contains(eventData.EventName))
+        {
+            return;
+        }
+
+        var payload = eventData.Payload[0] as IDictionary<string, object>;
+        var name = payload["Name"].ToString();
+        var isGauge = payload.ContainsKey("Mean");
+        Tuple<string, string> metricKey = new(eventData.EventSource.Name, name);
+
+        this.values[metricKey] = isGauge
+            ? Convert.ToDouble(payload["Mean"])
+            : Convert.ToDouble(payload["Increment"]) / this.options.RefreshIntervalSecs;
```

Comment:
The `Increment` is used for created an `CreateObservableCounter`. Is your suggestion still valid? It would also require a change to `EventCounterMetricsTests.cs:IncrementingPollingCounter` making the expected value 2 instead of 1.

Reply:
Its always possible I made a mistake, but I don't think that I did :) 

For example imagine that we have an IncrementingEventCounter, we call Increment() once per second, and the refresh interval reports an event back to us every 3 seconds:

| Time                 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |
| ---------------------|---|---|---|---|---|---|---|---|---|
| Increment() argument | 2 | 3 | 1 | 4 | 3 | 2 | 2 | 1 | 2 | 
| payload["Increment"] | - | - | 6 | - | - | 8 | - | - | 5 |

The ObservableCounter instrument expects that you provide the running total in the callback. So in this example you would need to provide 6 at t_3, 14 at t_6, and 19 t_9.